x/auth/account.go

```diff
@@ -122,12 +140,335 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+			j++
+		}
+		for k < len(bva.DelegatedVesting) && bva.DelegatedVesting[k].Denom != coin.Denom {
+			k++
+		}
+
+		baseAmt := coin.Amount
+
+		vestingAmt := sdk.ZeroInt()
+		if len(vestingCoins) > 0 {
+			vestingAmt = vestingCoins[j].Amount
+		}
+
+		delVestingAmt := sdk.ZeroInt()
+		if len(bva.DelegatedVesting) > 0 {
+			delVestingAmt = bva.DelegatedVesting[k].Amount
+		}
+
+		// compute min((BC + DV) - V, BC) per the specification
+		min := sdk.MinInt(baseAmt.Add(delVestingAmt).Sub(vestingAmt), baseAmt)
+		spendableCoin := sdk.NewCoin(coin.Denom, min)
+
+		if !spendableCoin.IsZero() {
+			spendableCoins = spendableCoins.AddCoinByDenom(spendableCoin)
+		}
+	}
+
+	return spendableCoins
+}
+
+// trackDelegation tracks a delegation amount for any given vesting account type
+// given the amount of coins currently vesting. It returns the resulting base
+// coins.
+//
+// CONTRACT: The account's coins, delegation coins, vesting coins, and delegated
+// vesting coins must be sorted.
+func (bva *BaseVestingAccount) trackDelegation(vestingCoins, amount sdk.Coins) {
+	bc := bva.GetCoins()
+
+	i, j, k := 0, 0, 0
+	for _, coin := range amount {
+		// zip/lineup all coins by their denomination to provide O(n) time
+		for i < len(bc) && bc[i].Denom != coin.Denom {
+			i++
+		}
+		for j < len(vestingCoins) && vestingCoins[j].Denom != coin.Denom {
+			j++
+		}
+		for k < len(bva.DelegatedVesting) && bva.DelegatedVesting[k].Denom != coin.Denom {
+			k++
+		}
+
+		baseAmt := sdk.ZeroInt()
+		if len(bc) > 0 {
+			baseAmt = bc[i].Amount
+		}
+
+		vestingAmt := sdk.ZeroInt()
+		if len(vestingCoins) > 0 {
+			vestingAmt = vestingCoins[j].Amount
+		}
+
+		delVestingAmt := sdk.ZeroInt()
+		if len(bva.DelegatedVesting) > 0 {
+			delVestingAmt = bva.DelegatedVesting[k].Amount
+		}
+
+		// Panic if the delegation amount is zero or if the base coins does not
+		// exceed the desired delegation amount.
+		if coin.Amount.IsZero() || baseAmt.LT(coin.Amount) {
+			panic("delegation attempt with zero coins or insufficient funds")
+		}
+
+		// compute x and y per the specification, where:
+		// X := min(max(V - DV, 0), D)
+		// Y := D - X
+		x := sdk.MinInt(sdk.MaxInt(vestingAmt.Sub(delVestingAmt), sdk.ZeroInt()), coin.Amount)
+		y := coin.Amount.Sub(x)
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedVesting = bva.DelegatedVesting.AddCoinByDenom(xCoin)
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedFree = bva.DelegatedFree.AddCoinByDenom(yCoin)
+		}
+
+		bva.Coins = bva.Coins.SubCoinByDenom(coin)
+	}
+}
+
+// TrackUndelegation tracks an undelegation amount by setting the necessary
+// values by which delegated vesting and delegated vesting need to decrease and
+// by which amount the base coins need to increase. The resulting base coins are
+// returned.
+//
+// CONTRACT: The account's coins and undelegation coins must be sorted.
+func (bva *BaseVestingAccount) TrackUndelegation(amount sdk.Coins) {
```

Comment:
ditto

Reply:
we're only checking if ≠ 0

Reply:
I don't really think there is a need to? Are you asking to perform a `amount.IsValid()` call here? We know the coins cannot be negative and we already check the `0` case. So why incur the overhead?